App/Services/RpcController.cs

```diff
@@ -0,0 +1,248 @@
+    /// <exception>Throws an exception if reconnection fails. Exceptions from disconnection are ignored.</exception>
+    public Task Reconnect(CancellationToken ct = default);
+
+    /// <summary>
+    ///     Start the VPN using the stored credentials in the ICredentialManager. If the VPN is already running, this
+    ///     may have no effect.
+    /// </summary>
+    /// <exception cref="InvalidOperationException">Another operation is in progress</exception>
+    /// <exception cref="RpcOperationException">If the sending of the start command fails</exception>
+    /// <exception cref="VpnLifecycleException">If the service reports that the VPN failed to start</exception>
+    public Task StartVpn(CancellationToken ct = default);
+
+    /// <summary>
+    ///     Stop the VPN. If the VPN is already not running, this may have no effect.
+    /// </summary>
+    /// <exception cref="InvalidOperationException">Another operation is in progress</exception>
+    /// <exception cref="RpcOperationException">If the sending of the stop command fails</exception>
+    /// <exception cref="VpnLifecycleException">If the service reports that the VPN failed to stop</exception>
+    public Task StopVpn(CancellationToken ct = default);
+}
+
+public class RpcController : IRpcController
+{
+    private readonly ICredentialManager _credentialManager;
+
+    private readonly RaiiSemaphoreSlim _operationLock = new(1, 1);
+    private Speaker<ClientMessage, ServiceMessage>? _speaker;
+
+    private readonly RaiiSemaphoreSlim _stateLock = new(1, 1);
+    private readonly RpcModel _state = new();
+
+    public RpcController(ICredentialManager credentialManager)
+    {
+        _credentialManager = credentialManager;
+    }
+
+    public event EventHandler<RpcModel>? StateChanged;
+
+    public RpcModel GetState()
+    {
+        using var _ = _stateLock.Lock();
+        return _state.Clone();
+    }
+
+    public async Task Reconnect(CancellationToken ct = default)
+    {
+        using var _ = await AcquireOperationLockNowAsync();
+        MutateState(state =>
+        {
+            state.RpcLifecycle = RpcLifecycle.Connecting;
+            state.VpnLifecycle = VpnLifecycle.Stopped;
+            state.VisibleAgents.Clear();
+        });
+
+        if (_speaker != null)
+            try
+            {
+                await _speaker.DisposeAsync();
+            }
+            catch (Exception e)
+            {
+                // TODO: log/notify?
+                Debug.WriteLine($"Error disposing existing Speaker: {e}");
+            }
+
+        try
+        {
+            var client =
+                new NamedPipeClientStream(".", "Coder.Desktop.Vpn", PipeDirection.InOut, PipeOptions.Asynchronous);
+            await client.ConnectAsync(ct);
+            _speaker = new Speaker<ClientMessage, ServiceMessage>(client);
+            _speaker.Receive += SpeakerOnReceive;
+            _speaker.Error += SpeakerOnError;
```

Comment:
We add ourselves to the event callback for the speaker here, do we need to remove ourselves around line 105 when we discard the old speaker?

Reply:
We don't need to since it's getting destroyed but I'll add it anyways